architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,446 @@
+- With this work, users will be able to deploy Radius resources using GitOps tools by authoring Kubernetes CRDs that contain Bicep manifests. We will essentially be providing a "translation layer" between Kubernetes resources (that Flux and ArgoCD can understand) and Radius resources (that Radius can understand).
+
+**Goal: Users can deploy Radius resources using GitOps tools**
+- We will be supporting the full suite of Radius resources in the BicepDeployment resource, so users can deploy any Radius resource using GitOps tools.
+
+**Goal: Users can enable this feature without writing code of their own**
+- We will provide a CLI command that generates the BicepDeployment resource from a Bicep manifest and parameters file, so users can enable this feature without writing code of their own.
+
+### Non-goals
+
+<!--
+Describe non-goals to identify something that we won’t be focusing on 
+immediately. We won’t be expending any effort on these matters. If there
+will be follow-ups after this work, list them here. If there are things
+we plan to do in the future, but are out of scope of this design, list
+them here. Provide a brief explanation on why this is a non-goal.
+-->
+
+**Non-goal: Full support for GitOps**
+- We will not yet be implementing automatic generation of BicepDeployment resources from Bicep manifests or querying Git repositories. This design will be covered in a separate design document.
+
+### User scenarios
+
+<!--
+Describe the user scenarios for this design. Ensure that you define the
+roles and personas in these user scenarios when it requires API design.
+If you have an existing issue that describes the user scenarios, please
+link to that issue instead.
+-->
+
+#### Jon can use GitOps to deploy a standard Radius environment for developer use
+
+Jon is an infrastructure operator who uses GitOps to automate the creation of developer Kubernetes clusters. Today, he publishes Kubernetes manifests to a GitHub repository, and a GitOps tool watches the repository and applies the manifests to the Kubernetes cluster. He wants to have a Radius environment and a set of recipes available on the clusters for his developers to use, since they are building applications using Radius.
+
+He first sets up the Radius Helm chart as a dependency in the cluster, and then runs the `rad bicep generate bicepdeployment env.bicep --parameters env.bicepparam` command on his machine to create an BicepDeployment resource that contains the template and parameters specified. He then commits the BicepDeployment resource to the Git repository.
+
+When a developer creates a new development cluster using the configuration specified in Jon's GitHub repository with a GitOps tool, the BicepDeployment resource is applied on the cluster, and the Radius environment defined in the `env.bicep` Bicep manifest is deployed. The developer can then use the environment to deploy and test their applications using Radius.
+
```

Comment:
let's add a non-gitops scenario, where a user can just use Radius to deploy Bicep resources using this feature (Azure resources?)

Reply:
In fact, we can probably remove the gitops-specific language here